tests/test.c

```diff
@@ -24,134 +24,50 @@
+    static char *opt = opt_tests;
 
-static char memmove_string[4];
-static char range_string[] = "123456";
-static char *src, *dst;
+    *out_name = strtok(opt, opt_test_delims);
 
-static void cpu_freq_expect(const char *cpu_str, uint64_t expectation) {
-    uint64_t result = get_cpu_freq(cpu_str);
-    if (result != expectation) {
-        printk("Could not parse cpu frequency from string '%s'\n", cpu_str);
-        printk("Expectation vs. result: %llu vs. %llu\n", expectation, result);
-        BUG();
+    opt = NULL;
+    if (*out_name) {
+        *out_test_fn = symbol_address(*out_name);
+        if (!*out_test_fn)
+            panic("Symbol for test %s not found\n", *out_name);
+        return 1;
     }
-
-    printk("Got CPU string '%s' and frequency '%llu'\n", cpu_str, result);
-    return;
+    return 0;
 }
 
-static int __user_text func(void *arg) { return 0; }
-
-static int ktf_unit_tests(void) {
-    usermode_call(func, NULL);
-
-    printk("\nLet the UNITTESTs begin\n");
-    printk("Commandline parsing: %s\n", kernel_cmdline);
-
-    if (strcmp(opt_string, "foo")) {
-        printk("String parameter opt_string != foo: %s\n", opt_string);
-        BUG();
-    }
-    else {
-        printk("String parameter parsing works!\n");
-    }
-
-    if (strcmp(opt_badstring, "tool")) {
-        printk("String parameter opt_badstring != tool: %s\n", opt_badstring);
-        BUG();
-    }
-    else {
-        printk("String parameter parsing works!\n");
-    }
-
-    if (opt_ulong != 42) {
-        printk("Integer parameter opt_ulong != 42: %d\n", opt_ulong);
-        BUG();
-    }
-    else {
-        printk("Integer parameter parsing works!\n");
-    }
+void test_main(void) {
+    char *name;
+    test_fn *fn = NULL;
+    unsigned n = 0;
 
-    if (!opt_bool || !opt_booltwo) {
-        printk("Boolean parameter opt_bool != true: %d\n", opt_bool);
-        printk("Boolean parameter opt_booltwo != true: %d\n", opt_booltwo);
-        BUG();
-    }
-    else {
-        printk("Boolean parameter parsing works!\n");
-    }
+    printk("\nRunning tests\n");
 
-    printk("\nMemmove testing:\n");
-    (void) memmove(memmove_string, opt_string, sizeof(opt_string));
-    if (!strcmp(memmove_string, opt_string)) {
-        printk("Moving around memory works!\n");
-    }
-    else {
-        printk("Memmove'ing did not work: %s (%p) != %s (%p)\n", memmove_string,
-               memmove_string, opt_string, opt_string);
-    }
+    while (get_next_test(&fn, &name)) {
+        int rc;
 
-    src = (char *) range_string;
-    dst = (char *) range_string + 2;
-    (void) memmove(dst, src, 4);
-    if (!strcmp(range_string, "121234")) {
-        printk("Moving around memory with overlaping ranges works!\n");
-    }
-    else {
-        printk("Overlaping memmove'ing did not work: %s != %s\n", range_string, "121234");
+        printk("Running test: %s\n", name);
+        rc = fn(NULL);
+        if (rc)
+            panic("Test returned: 0x%x\n", rc);
```

Comment:
please add name of the test here

Reply:
Please do not panic on error

Reply:
Maybe always print the return value for each test regardless of return value - this highlights when a test finishes